File: basic_dsa_one_hundred_twenty_three.cpp
/**
 * Basic DSA - Invert Binary Tree (LeetCode 226)
 * Invert the tree (swap left and right subtrees).
 */
#include <iostream>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* invertTree(TreeNode* root) {
    if (!root) return nullptr;
    TreeNode* tmp = root->left;
    root->left = invertTree(root->right);
    root->right = invertTree(tmp);
    return root;
}

void printInorder(TreeNode* n) {
    if (!n) return;
    printInorder(n->left);
    std::cout << n->val << " ";
    printInorder(n->right);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Invert Binary Tree\n";
    std::cout << std::string(60, '=') << "\n";

    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(7);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(9);

    invertTree(root);
    std::cout << "[4,2,7,1,3,6,9] inverted inorder -> ";
    printInorder(root);
    std::cout << "\n";
}

File: basic_dsa_one_hundred_twenty_three.py
"""
Basic DSA - Find Peak Element
Given an array nums, find a peak element and return its index.
A peak element is an element that is strictly greater than its neighbors.
You may assume nums[-1] = nums[n] = -inf.
"""


def find_peak_element(nums):
    """
    Binary search on slope.
    Time Complexity: O(log n)
    Space Complexity: O(1)
    """
    if not nums:
        return None

    left, right = 0, len(nums) - 1
    while left < right:
        mid = left + (right - left) // 2
        if nums[mid] < nums[mid + 1]:
            left = mid + 1
        else:
            right = mid
    return left


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Find Peak Element")
    print("=" * 60)

    tests = [
        ([1, 2, 3, 1], 2),         # peak at index 2 (value 3)
        ([1, 2, 1, 3, 5, 6, 4], 5),# one valid peak index is 5 (value 6)
        ([1], 0),
        ([2, 1], 0),
        ([1, 2], 1),
    ]

    for nums, expected in tests:
        got = find_peak_element(nums)
        print(f"nums={nums} -> peak_index={got} (expected {expected})")

